src/sentry/snuba/metrics/datasource.py

```diff
@@ -420,6 +421,138 @@ def get_tag_values(
     return tags
 
 
+@dataclass
+class GroupLimitFilters:
+    """Fields and values to filter queries when exceeding the Snuba query limit.
+
+    Snuba imposes a limit on the number of rows that can be queried and
+    returned. This limit can be exceeded when grouping metrics by one or more
+    tags. In this case, we take the first groups returned by Snuba and filter
+    subsequent queries with this set of tag values.
+
+    Fields:
+
+    - ``keys``: A tuple containing resolved tag names ("tag[123]") in the order
+      of the ``groupBy`` clause.
+
+    - ``values``: A list of tuples containing the tag values of the group keys.
+      The list is in the order returned by Snuba. The tuple elements are ordered
+      like ``keys``.
+
+    - ``conditions``: A list of raw snuba query conditions to filter subsequent
+      queries by.
+    """
+
+    keys: Tuple[Groupable]
+    values: List[Tuple[int]]
+    conditions: ConditionGroup
+
+
+def _get_group_limit_filters(
+    query: QueryDefinition, results: List[Mapping[str, int]]
+) -> Optional[GroupLimitFilters]:
+    if not query.groupby or not results:
+        return None
+
+    # Translate the groupby fields of the query into their tag keys because these fields
+    # will be used to filter down and order the results of the 2nd query.
+    # For example, (project_id, transaction) is translated to (project_id, tags[3])
+    keys = tuple(
+        resolve_tag_key(query.org_id, field) if field not in ALLOWED_GROUPBY_COLUMNS else field
+        for field in query.groupby
+    )
+
+    values = [tuple(row[col] for col in keys) for row in results]
```

Comment:
I think we should remove duplicates here, because `results` could be a from a time series query, where groups appear once for every time slot.